### PR TITLE
irq: introduce new `irq_set_affinity()` API - alternate

### LINKED ISSUE #78138: riscv: plic: IRQs may not be enabled in non-zero HART
**Is your enhancement proposal related to a problem? Please describe.**

All enabled HARTs should be able to handle the IRQ. #65922 is supposed to enable the IRQ on all HARTs, however the patch assumed that the first HART of a SoC has 1 context, whereas the other HARTs have 2 contexts each, i.e.:

| HART | CONTEXT(s) |
|:-:|:-:|
| 0 | M |
| 1 | M,S |
| 2 | M,S |
| N | M,S |

https://github.com/zephyrproject-rtos/zephyr/blob/5b15751d0bcb55afc8074c38893b14b0bba98a75/drivers/interrupt_controller/intc_plic.c#L112-L134

This means that for SoC that doesn't fit into that assumption, the implementation will not work, i.e. QEMU RISC-V SoCs, which has 2 contexts per HART:
https://github.com/zephyrproject-rtos/zephyr/blob/5b15751d0bcb55afc8074c38893b14b0bba98a75/dts/riscv/qemu/virt-riscv.dtsi#L164-L182

| HART | CONTEXT(s) |
|:-:|:-:|
| 0 | M(11),S(9) |
| 1 | M(11),S(9) |
| 2 | M(11),S(9) |
| N | M(11),S(9) |

The current PLIC driver will enable the interrupt on context 0, 1, 3, .., (hartid * 2 - 1) of the QEMU SoC, which corresponds to:

| HART | CONTEXT(s) |
|:-:|:-:|
| 0 | M |
| 1 | S |
| 2 | S |
| N | S |

as a result, machine external interrupts, like UART, only works on `hartid=0`

This can be tested by building the `hello_world` with `SHELL` enabled (to test the UART interrupt), but boot it on `hartid=1` instead of `0`:

```
west build -b qemu_riscv64 -p auto -t run zephyr/samples/hello_world -- \
  -DCONFIG_SHELL=y \
  -DCONFIG_RV_BOOT_HART=1 \
  -DCONFIG_MP_MAX_NUM_CPUS=2
```

Since UART IRQ isn't enabled on the context M of HART 1, there will be no prompt on the terminal, and typing in the terminal will have no response from the device:

```
*** Booting Zephyr OS build v3.7.0-2380-g5b15751d0bcb ***
Hello World! qemu_riscv64/qemu_virt_riscv64
```

The workaround for QEMU RISC-V SoCs is simple:

```diff
--- a/drivers/interrupt_controller/intc_plic.c
+++ b/drivers/interrupt_controller/intc_plic.c
@@ -111,7 +111,7 @@ static inline uint32_t get_plic_enabled_size(const struct device *dev)
 
 static inline uint32_t get_first_context(uint32_t hartid)
 {
-       return hartid == 0 ? 0 : (hartid * 2) - 1;
+       return hartid * 2;
 }
 
 static inline mem_addr_t get_context_en_addr(const struct device *dev, uint32_t cpu_num)
```

Run the same build command again, the terminal will be responsive this time around:

```
*** Booting Zephyr OS build v3.7.0-2380-g5b15751d0bcb ***
Hello World! qemu_riscv64/qemu_virt_riscv64

uart:~$ blahblahblah
```

However, this workaround only works on SoC where each hart has 2 contexts.

**Describe the solution you'd like**

Update the implementation so that the interrupt will work regardless of the mapping of the contexts.

**Describe alternatives you've considered**

Not fixing it.

### PR DESCRIPTION
**Pull request by ycsin**
> [!NOTE]
> This implementation doesn't require empty stubs in all of the architectures, at the expense of an additional precompiler switch in the `irq.h` header
>
> This is an alternate implementation of #77645 that creates empty stubs for each of the architectures, check that PR out for more info.

`irq_set_affinity(irq, mask)` allows the runtime affinity configuration of an interrupt line, which can be helpful in tuning the performance in a SMP setup (i.e. `irqbalance`).

An implementation is added for RISC-V PLIC in this PR.

___

# Kconfig dependencies

image

___

# Testing on `qemu_riscv64_smp`

```
west build -b qemu_riscv64_smp -p always -t run zephyr/samples/hello_world -- \
  -DCONFIG_SHELL=y \
  -DCONFIG_MP_MAX_NUM_CPUS=4 \
  -DCONFIG_IRQ_AFFINITY=y \
  -DCONFIG_PLIC_SHELL=y
```

```
[125/126] To exit from QEMU enter: 'CTRL+a, x'[QEMU] CPU: riscv64
*** Booting Zephyr OS build v3.7.0-1894-g2555dd2bb4f6 ***
Hello World! qemu_riscv64/qemu_virt_riscv64/smp

uart:~$ plic stats get interrupt-controller@c000000 
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         105           0           0           0         123      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         211           0           0           0         229      0x800058fa(0x8000a130)

uart:~$ plic affinity set interrupt-controller@c000000 10 0x2
IRQ 10 affinity set to 0x2
uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         110           0           0         490      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         208           0           0         589      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         315           0           0         695      0x800058fa(0x8000a130)

uart:~$ plic affinity set interrupt-controller@c000000 10 0x4
IRQ 10 affinity set to 0x4
uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         404         100           0         880      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         404         206           0         986      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         404         305           0        1085      0x800058fa(0x8000a130)

uart:~$ plic affinity set interrupt-controller@c000000 10 0x8
IRQ 10 affinity set to 0x8
uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         404         398         114        1287      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
    10         367         404         398         216        1390      0x800058fa(0x8000a130)

uart:~$ plic affinity set interrupt-controller@c000000 10 0xf
IRQ 10 affinity set to 0xF
uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
     0          12          16          14          20          71      0x800045f4(0)
    10         391         432         421         345        1598      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
     0          38          38          40          40         162      0x800045f4(0)
    10         417         457         452         371        1706      0x800058fa(0x8000a130)

uart:~$ plic stats get interrupt-controller@c000000
   IRQ       CPU 0       CPU 1       CPU 2       CPU 3       Total      ISR(ARG)
     0          62          60          65          58         252      0x800045f4(0)
    10         438         484         486         398        1815      0x800058fa(0x8000a130)

uart:~$ 
```

### CODE REVIEW

File: drivers/interrupt_controller/intc_plic.c
Comment (andyross) [inline drivers/interrupt_controller/intc_plic.c:79]: The API only specifies a uint32_t argument for the mask, so seems needless to try to support 64 bit here.  Or else we should widen the API.

**Review by andyross — DISMISSED** [review]
Seems clean.  This version with the kconfigs stubbing gets my vote for sure.  One nitpick about mismatched bit widths in the riscv code.

File: drivers/interrupt_controller/Kconfig.plic
Comment (cfriedt) [inline drivers/interrupt_controller/Kconfig.plic:28]: I kind of wonder whether the default mask should be 0xffffffff, which would (presumably) allow irqs to be handled by the first available core in an SMP configuration.

Another random thought - it would be kind of ideal if the implementation would also unpend the IRQ on other cores after "winning the race" (if that isn't done at the hardware level, and the arch has the capability to do so in hw).

Additionally, why not make this mask option arch-independent / generic?

Comment (ycsin) [reply drivers/interrupt_controller/Kconfig.plic:28]: > I kind of wonder whether the default mask should be 0xffffffff, which would (presumably) allow irqs to be handled by the first available core in an SMP configuration.

That's debatable, I left the default as `0x1` as that's the default behavior when IRQ affinity isn't enabled

> Another random thought - it would be kind of ideal if the implementation would also unpend the IRQ on other cores after "winning the race" (if that isn't done at the hardware level, and the arch has the capability to do so in hw).

On RISCV, there seems to be no way for the core that successfully claimed that IRQ to stop the PLIC from waking up other cores.

RISCV Interrupt Notification

> Additionally, why not make this mask option arch-independent / generic?

Makes sense, will do that soon

Comment (cfriedt) [reply drivers/interrupt_controller/Kconfig.plic:28]: > That's debatable, I left the default as `0x1` as that's the default behavior when IRQ affinity isn't enabled

Is 0x1 the default for all architectures or just RISC-V?

Future thing to consider though. I don't think it's a big deal.

Comment (ycsin) [reply drivers/interrupt_controller/Kconfig.plic:28]: > > That's debatable, I left the default as `0x1` as that's the default behavior when IRQ affinity isn't enabled
> 
> Is 0x1 the default for all architectures or just RISC-V?

It will be the default value for all arch when I make the Kconfig generic

Comment (cfriedt) [reply drivers/interrupt_controller/Kconfig.plic:28]: I guess my question is whether all architectures currently only allow a single core in a cluster to receive interrupts.

Comment (ycsin) [reply drivers/interrupt_controller/Kconfig.plic:28]: > I guess my question is whether all architectures currently only allow a single core in a cluster to receive interrupts.

Sorry, no idea about that yet 😬, hopefully other arch maintainers can answer

Comment (ycsin) [reply drivers/interrupt_controller/Kconfig.plic:27]: I've moved that Kconfig to `arch/Kconfig` and removed the default value, so each implementation is free to define their own default, or leave it to the user when they set `CONFIG_IRQ_AFFINITY=y`.

The default value for RISC-V is defined to be `0x1` in `Kconfig.plic`

File: drivers/interrupt_controller/Kconfig.plic
Comment (evgeniy-paltsev) [inline drivers/interrupt_controller/Kconfig.plic:25]: Is it expected that we have two `config IRQ_AFFINITY_MASK` entries? (first in arch/Kconfig)

Comment (ycsin) [reply drivers/interrupt_controller/Kconfig.plic:25]: Yes, an implementation can have a default value by overridding the one in `arch/Kconfig`

File: drivers/interrupt_controller/intc_plic.c
Comment (ycsin) [reply drivers/interrupt_controller/intc_plic.c:390]: The `mask` is following the convention of `irq`, I suggest that we do it in another PR before or after this PR to aligned the type of `irq` to `uint32_t` altogether, what do you think?

File: drivers/interrupt_controller/intc_plic.c
Comment (cfriedt) [inline drivers/interrupt_controller/intc_plic.c:608]: The indeces into this array are fairly confusing but this line can be simplified slightly using a previously declared variable. 

```suggestion
				      stat.irq_count[offset + i]);
```

In the `plic_stats` structure. Is `irq_count` a pointer to an array that is 2+ dimensions in size? Or does some other dimension reflect values other than irq counts?

It might be good in a future PR to clean that up a bit.

Comment (cfriedt) [reply drivers/interrupt_controller/intc_plic.c:608]: Reviewed #78140. I think it's a bit clearer now.  Is this an interleaved array based on hart contexts?

Comment (ycsin) [reply drivers/interrupt_controller/intc_plic.c:608]: For every instance of a PLIC, if `PLIC_SHELL` is enabled, there will be a `local_irq_count[NUM_CPUS][NUM_IRQS]`, which is pointed to by `stat.irq_count`.

I can probably make a (per instance) helper inline function to access this so that it is clearer

**Comment by ycsin** [comment]
> am I missing something?

Sorry, I forgot to update the build command - you need to add `CONFIG_IRQ_AFFINITY=y`

**Comment by con-pax** [comment]
> Sorry, I forgot to update the build command - you need to add `CONFIG_IRQ_AFFINITY=y`

Thanks! I though that and tried it however I got a bit of vomit: 

```shell
$ west build -p -b mpfs_icicle/polarfire/u54/smp zephyr/samples/hello_world/ -DCONFIG_SHELL=y -DCONFIG_PLIC_SHELL=y -DCONFIG_PLIC_IRQ_AFFINITY=y
-- west build: making build dir /mnt/dev/z_workspace/build pristine
-- west build: generating a build system
Loading Zephyr default modules (Zephyr base).
-- Application: /mnt/dev/z_workspace/zephyr/samples/hello_world
-- CMake version: 3.28.0-rc5
-- Found Python3: /usr/bin/python3 (found suitable version "3.8.10", minimum required is "3.8") found components: Interpreter 
-- Cache files will be written to: /home/conor/.cache/zephyr
-- Zephyr version: 3.7.99 (/mnt/dev/z_workspace/zephyr)
-- Found west (found suitable version "1.2.0", minimum required is "0.14.0")
-- Board: mpfs_icicle, qualifiers: polarfire/u54/smp
-- ZEPHYR_TOOLCHAIN_VARIANT not set, trying to locate Zephyr SDK
-- Found host-tools: zephyr 0.16.8 (/home/conor/zephyr-sdk-0.16.8)
-- Found toolchain: zephyr 0.16.8 (/home/conor/zephyr-sdk-0.16.8)
-- Found Dtc: /home/conor/zephyr-sdk-0.16.8/sysroots/x86_64-pokysdk-linux/usr/bin/dtc (found suitable version "1.6.0", minimum required is "1.4.6") 
-- Found BOARD.dts: /mnt/dev/z_workspace/zephyr/boards/microchip/mpfs_icicle/mpfs_icicle_polarfire_u54_smp.dts
-- Generated zephyr.dts: /mnt/dev/z_workspace/build/zephyr/zephyr.dts
-- Generated devicetree_generated.h: /mnt/dev/z_workspace/build/zephyr/include/generated/zephyr/devicetree_generated.h
-- Including generated dts.cmake file: /mnt/dev/z_workspace/build/zephyr/dts.cmake
Parsing /mnt/dev/z_workspace/zephyr/Kconfig
Loaded configuration '/mnt/dev/z_workspace/zephyr/boards/microchip/mpfs_icicle/mpfs_icicle_polarfire_u54_smp_defconfig'
Merged configuration '/mnt/dev/z_workspace/zephyr/samples/hello_world/prj.conf'
Merged configuration '/mnt/dev/z_workspace/build/zephyr/misc/generated/extra_kconfig_options.conf'

error: PLIC_IRQ_AFFINITY (defined at drivers/interrupt_controller/Kconfig.plic:17) is assigned in a
configuration file, but is not directly user-configurable (has no prompt). It gets its value
indirectly from other symbols. See
http://docs.zephyrproject.org/latest/kconfig.html#CONFIG_PLIC_IRQ_AFFINITY and/or look up
PLIC_IRQ_AFFINITY in the menuconfig/guiconfig interface. The Application Development Primer, Setting
Configuration Values, and Kconfig - Tips and Best Practices sections of the manual might be helpful
too.

CMake Error at /mnt/dev/z_workspace/zephyr/cmake/modules/kconfig.cmake:389 (message):
  command failed with return code: 1
Call Stack (most recent call first):
  /mnt/dev/z_workspace/zephyr/cmake/modules/zephyr_default.cmake:133 (include)
  /mnt/dev/z_workspace/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:66 (include)
  /mnt/dev/z_workspace/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:92 (include_boilerplate)
  CMakeLists.txt:5 (find_package)
```

**Comment by con-pax** [comment]
Whoops! my bad, CONFIG_IRQ_AFFINITY! 

```shell
*** Booting Zephyr OS build v3.7.0-2739-gb257c44d6f56 ***
Hello World! mpfs_icicle/polarfire/u54/smp

uart:~$ plic stats get interrupt-controller@c000000
   IRQ  CPU 0  CPU 1  CPU 2  CPU 3  Total       ISR(ARG)
    91      0     38      0      0     41       0x80005c6a(0x8000a4f8)

uart:~$ plic affinity set interrupt-controller@c000000 91 0x2
IRQ 91 affinity set to 0x2
uart:~$ plic stats get interrupt-controller@c000000
   IRQ  CPU 0  CPU 1  CPU 2  CPU 3  Total       ISR(ARG)
    91      0     87    433      0    522       0x80005c6a(0x8000a4f8)

uart:~$ plic affinity set interrupt-controller@c000000 91 0x4
IRQ 91 affinity set to 0x4
uart:~$ plic stats get interrupt-controller@c000000
   IRQ  CPU 0  CPU 1  CPU 2  CPU 3  Total       ISR(ARG)
    91      0     87    471     39    598       0x80005c6a(0x8000a4f8)

uart:~$ 
```

This looks really cool to me!

**Comment by con-pax** [comment]
I'm not around for a couple of weeks, but this deffo gets my vote

**Comment by ycsin** [comment]
> I'm not around for a couple of weeks, but this deffo gets my vote

Have a good one! 🏖️

**Comment by ycsin** [comment]
Rebased after ##78140

**Review by bjarki-andreasen — APPROVED** [review]
Looks good :)

**Comment by ycsin** [comment]
Having looked at a few more interrupt controller drivers and the (custom) APIs they implement, I'm not sure if this PR is the best path forward - the configuration of IRQ affinity should not be something special when there are custom APIs to set/clear pending and so on. Creating a new arch-level API might not be necessary since PLIC will be the only controller using it for the foreseeable future, and does little to the ad-hoc nature of our interrupt controller implementation.

I'm going to spin off the PLIC IRQ affinity implementation into a separate PR that can be reviewed separately from the introduction of `irq_set_affinity()` - #78681

Comments welcomed.

**Pull request closed without merging**
